File: libint-sys/src/bridge/shell.rs
use cxx::UniquePtr;

#[cxx::bridge(namespace = "libint2")]
pub mod ffi {

    unsafe extern "C++" {
        include!("libint-sys/shim/include/shell.h");

        // -----------------------------------------------------
        // Shell
        // -----------------------------------------------------
        type Shell;

        #[must_use]
        fn shell() -> UniquePtr<Shell>;

        #[must_use]
        fn alpha(shell: &Shell) -> Vec<f64>;
        #[must_use]
        fn contr(shell: &Shell) -> *const *const Contraction;
        #[must_use]
        fn O(shell: &Shell) -> [f64; 3];
        #[must_use]
        fn max_ln_coeff(shell: &Shell) -> Vec<f64>;

        #[Self = "Shell"]
        #[must_use]
        fn am_symbol(l: usize) -> c_char;
        #[Self = "Shell"]
        #[must_use]
        fn am_symbol_to_l(am_symbol: c_char) -> u16;

        // -----------------------------------------------------
        // Contraction
        // -----------------------------------------------------
        type Contraction;

        #[must_use]
        fn contraction() -> UniquePtr<Contraction>;

        // getters
        #[must_use]
        fn l(contraction: &Contraction) -> i32;
        #[must_use]
        fn pure(contraction: &Contraction) -> bool;
        #[must_use]
        fn coeff(contraction: &Contraction) -> Vec<f64>;

        #[must_use]
        fn cartesian_size(contraction: &Contraction) -> usize;
        #[must_use]
        fn size(contraction: &Contraction) -> usize;

    }
}

#[cfg(test)]
mod tests {
    use super::ffi::Shell;

    #[test]
    fn am_symbol() {
        assert_eq!(Shell::am_symbol(0), 115); // s
        assert_eq!(Shell::am_symbol(1), 112); // p
    }

    #[test]
    fn am_symbol_to_l() {
        assert_eq!(Shell::am_symbol_to_l(115), 0); // s
        assert_eq!(Shell::am_symbol_to_l(112), 1); // p
    }
}

File: libint-sys/shim/src/shell.cc
#include "libint-sys/shim/include/shell.h"
#include <iterator>
#include <memory>

namespace libint2 {
  // -----------------------------------------------------
  // Shell
  // -----------------------------------------------------
  std::unique_ptr<Shell> shell() {
    return std::make_unique<Shell>();
  }

  rust::Vec<double> alpha(const Shell& shell) {
    rust::Vec<double> v;
    std::copy(shell.alpha.begin(), shell.alpha.end(), std::back_inserter(v));
    return v;
  }

  const Contraction& at_contraction(const Shell& shell, std::size_t i) {
    return shell.contr[i];
  }

  std::array<double, 3> O(const Shell& shell) {
    return shell.O;
  }

  rust::Vec<double> max_ln_coeff(const Shell& shell) {
    rust::Vec<double> v;
    std::copy(shell.max_ln_coeff.begin(), shell.max_ln_coeff.end(), std::back_inserter(v));
    return v;
  }

  std::size_t cartesian_size(const Shell& shell) {
    return shell.cartesian_size();
  }
  std::size_t size(const Shell& shell) {
    return shell.size();
  }
  std::size_t ncontr(const Shell& shell) {
    return shell.ncontr();
  }
  std::size_t nprim(const Shell& shell) {
    return shell.nprim();
  }

  // -----------------------------------------------------
  // Contraction
  // -----------------------------------------------------

  // getters
  
  int l(const Contraction& contraction) {
    return contraction.l;
  }

  bool pure(const Contraction& contraction) {
    return contraction.pure;
  }
  
  rust::Vec<double> coeff(const Contraction& contraction) {
    rust::Vec<double> v;
    std::copy(contraction.coeff.begin(), contraction.coeff.end(), std::back_inserter(v));
    return v;
  }

  size_t cartesian_size(const Contraction& contraction) {
    return contraction.cartesian_size();
  }

  size_t size(const Contraction& contraction) {
    return contraction.size();
  }
}
